e2e/rapid_test.go

```diff
@@ -0,0 +1,579 @@
+
+func TestCheckLivenessIssueCheck_Fails(t *testing.T) {
+	numOfNodes := 4
+	rounds := map[uint64]map[int][]int{
+		0: {
+			0: {0, 1, 3, 2},
+			1: {0, 1, 3},
+			2: {0, 1, 2, 3},
+		},
+	}
+
+	countPrepare := 0
+	ft := &fakeTransport{
+		GossipFunc: func(ft *fakeTransport, msg *pbft.MessageReq) error {
+			routing, changed := rounds[msg.View.Round]
+			if changed {
+				//todo hack
+				from, err := strconv.Atoi(string(msg.From))
+				if err != nil {
+					t.Fatal(err)
+				}
+				for _, nodeId := range routing[from] {
+					// restrict prepare messages to node 3 in round 0
+					if msg.Type == pbft.MessageReq_Prepare && nodeId == 3 {
+						countPrepare++
+						if countPrepare == 3 {
+							fmt.Println("Ignoring prepare 3")
+							continue
+						}
+					}
+					// do not send commit for round 0
+					if msg.Type == pbft.MessageReq_Commit {
+						continue
+					}
+
+					ft.nodes[nodeId].PushMessage(msg)
+				}
+			} else {
+				for i := range ft.nodes {
+					from, _ := strconv.Atoi(string(msg.From))
+					// for rounds >0 do not send messages to/from node 3
+					if i == 3 || from == 3 {
+						continue
+					} else {
+						ft.nodes[i].PushMessage(msg)
+					}
+				}
+			}
+
+			return nil
+		},
+	}
+	cluster, timeoutsChan := generateCluster(numOfNodes, ft)
+	for i := range cluster {
+		cluster[i].SetInitialState(context.Background())
+	}
+
+	stuckList := NewBoolSlice(len(cluster))
+	doneList := NewBoolSlice(len(cluster))
+
+	callNumber := 0
+	for {
+		callNumber++
+		err := runClusterCycle(cluster, callNumber, stuckList, doneList)
+		if err != nil {
+			t.Error(err)
+		}
+
+		setDoneOnDoneState(cluster, doneList)
+		if stuckList.CalculateNum(true) == numOfNodes {
+			for i := range timeoutsChan {
+				timeoutsChan[i] <- time.Now()
+			}
+		}
+
+		//check that 3 node switched to done state
+		if doneList.CalculateNum(true) >= 3 {
+			return
+		}
+
+		if getMaxClusterRound(cluster) > 5 {
+			//todo remove comment when fix liveness issue
+			//t.Error("Infinite rounds")
+			return
+		}
+	}
+}
+
+func TestCheckLivenessIssue2Check(t *testing.T) {
+	numOfNodes := 5
+	rounds := map[uint64]map[int][]int{
+		0: {
+			0: {0, 1, 2},
+			1: {0},
+			2: {0, 3},
+			3: {0},
+		},
+		1: {
+			0: {1},
+			1: {1, 2, 3},
+			2: {1},
+			3: {1},
+		},
+	}
+
+	ft := &fakeTransport{
+		GossipFunc: func(ft *fakeTransport, msg *pbft.MessageReq) error {
+			routing, changed := rounds[msg.View.Round]
+			if changed {
+				from, err := strconv.Atoi(string(msg.From))
+				if err != nil {
+					t.Fatal(err)
+				}
+				for _, nodeId := range routing[from] {
+					ft.nodes[nodeId].PushMessage(msg)
+				}
+			} else {
+				for i := range ft.nodes {
+					ft.nodes[i].PushMessage(msg)
+				}
+			}
+			return nil
+		},
+	}
+	cluster, timeoutsChan := generateCluster(numOfNodes, ft)
+	for i := range cluster {
+		cluster[i].SetInitialState(context.Background())
+	}
+
+	stuckList := NewBoolSlice(len(cluster))
+	doneList := NewBoolSlice(len(cluster))
+
+	callNumber := 0
```

Comment:
Is there a way to abstract all this process? It seems to me that it is always doing the same thing:
- Get `stuck` and `done` list
- For loop over call number.
- Call `runClusterCycle`
- `setDoneOnDoneState` and do `timeoutsChan`.
- Check `doneList`.
- Check `getMaxClusterRound`.